### DISCUSSION TITLE
Fixing corrupted OGG container?

### DISCUSSION THREAD

**Discussion by Losses** [OptiVorbis/OptiVorbis#86] [Q&A] [answered]
I recently came across a Vorbis-encoded OGG file that seems to have undergone some non-standard processing, causing issues with recognition by `ffmpeg`. Interestingly, VLC can play the file without any problems. I attempted to repair the corrupted container format using OptiVorbis, but unfortunately, the attempt was unsuccessful.

```
[in#0 @ 0000020818a89f00] Error opening input: End of file
Error opening input file .\input.ogg.
Error opening input files: End of file
```

OptiVorbis utilizes Rust Audio's OGG crate for file parsing, but it appears that this crate strictly adheres to the RFC specifications and lacks robust fault tolerance (as we all know, fault tolerance in media encoding is a deep and complex issue). Specifically, I encountered an error while parsing the file:
[Link to the specific error in Rust Audio's OGG crate](https://github.com/RustAudio/ogg/blob/4b22c17d4ac365e8f4e16c69f0e906a95731e781/src/reading.rs#L50).

To the best of my knowledge, the only library currently capable of demuxing this audio file is Symphonia's OGG Demuxer. However, this library does not implement a muxer, which might pose challenges during the oggToOgg process:
[Link to Symphonia's Ogg Demuxer](https://github.com/pdeljanov/Symphonia/blob/master/symphonia-format-ogg/src/demuxer.rs).

Upon a cursory review of the Demuxer's implementation, it seems to deviate slightly from the RFC specifications.

If anyone is interested in this case, I can send the file via email for further analysis. I am also delving into the file's structure myself. While I have previously implemented my own OGG Demuxer, my expertise is more focused on Opus, and Vorbis presents an entirely different set of challenges, making this process quite arduous.

Any insights, suggestions, or assistance would be greatly appreciated!

**Comment by AlexTMjugador** [comment]
Hello, thanks for reaching out!

During the initial development of OptiVorbis, I found the [`oggz` tools](https://gitlab.xiph.org/xiph/liboggz) very useful for debugging low-level details about Ogg files. Since `oggz` uses `libogg` under the hood, I wouldn't expect it to decode files that other players based on `libogg` can't decode (VLC most likely isn't based on libogg). However, `oggz` might show some partial output or error messages that help narrow down what is wrong with the file.

I'm curious about your file and would be happy to receive it via email to potentially share some interesting observations. You can find my email address in the `Cargo.toml` file of this repository and on my website, which is linked from my GitHub profile. Please note that at the moment I'm quite busy with work and other commitments, so I can't guarantee any specific outcomes out of this, but I'll try to do my best to help.

**Reply by Losses** [reply, marked as answer]
> Hello, thanks for reaching out!
> 
> During the initial development of OptiVorbis, I found the [`oggz` tools](https://gitlab.xiph.org/xiph/liboggz) very useful for debugging low-level details about Ogg files. Since `oggz` uses `libogg` under the hood, I wouldn't expect it to decode files that other players based on `libogg` can't decode (VLC most likely isn't based on libogg). However, `oggz` might show some partial output or error messages that help narrow down what is wrong with the file.
> 
> I'm curious about your file and would be happy to receive it via email to potentially share some interesting observations. You can find my email address in the `Cargo.toml` file of this repository and on my website, which is linked from my GitHub profile. Please note that at the moment I'm quite busy with work and other commitments, so I can't guarantee any specific outcomes out of this, but I'll try to do my best to help.

Thank you for your prompt reply and interest in this matter!

I have conducted some personal research and found several answers regarding the issue at hand.

# Existing implementations

Here are some background information about existing codec implementations:

`libvorbis` strictly adheres to the specifications and has stringent requirements for file content. Specifically, it mandates that the first packet must be an `identification` header, the second a `comment` header, and the third a `setup` header. Any deviation from this order results in an error.

VLC and Symphonia do not enforce these requirements in their implementations. Symphonia, in particular, skips the parsing of the comment header and directly searches for the headers it is interested in, without regard to other constraints, making it highly fault-tolerant.

Interestingly, while VLC can handle header disorder, it still encounters issues during file conversion if there is erroneous information at the end of the file, but Symphonia do not have such troubles.

# The trouble

Returning to the issue I encountered, the file I have may contain an extra packet before the identification header, possibly added by the vendor for fingerprinting or some other special reason. This extra packet causes `libvorbis` to fail in decoding the file correctly. 

Additionally, the appended invalid information at the end of the file prevents VLC from processing it correctly. Based on the above research, I created a repair tool: [trimOggVorbisFile.ts](https://github.com/Web-Media-Foundation/infrastructure/blob/master/packages/ogg-polyfill/example/trimOggVorbisFile.ts).

In repairing OGG/Vorbis files, it might be necessary to **address both the container and the decoding frontend**, or at least **provide more detailed error information to help users understand what went wrong**. Currently, libraries in the Rust ecosystem generally lack consideration in this regard.

# Solution

Based on the above research, I attempted to create a repair tool.

However, the existing Vorbis codec implementations, including Lewton and Symphonia, do not expose the details of their OGG and Vorbis decoding frontends. Additionally, I am not a Rust developer. Therefore, I implemented an OGG decoding frontend from scratch in TypeScript: [OggVorbisPage.ts](https://github.com/Web-Media-Foundation/infrastructure/blob/master/packages/ogg-polyfill/src/OggVorbisPage.ts), with some packet editing feature.

This decoding frontend is designed to parse metadata; it is debugging-oriented code rather than production-oriented, strictly adhering to the specifications without any shortcuts, and provides more detailed error handling, including the detail about the Vorbis spec.

This debugging frontend also exposes all intermediate information generated during the decoding process, helping developers better debug their files. Here is a demonstration of its usage: [fetchOggVorbisFile.ts](https://github.com/Web-Media-Foundation/infrastructure/blob/master/packages/ogg-polyfill/src/fetchOggVorbisFile.ts).

Thank you again for your attention and willingness to help. I hope this information proves useful, and I look forward to any further insights or suggestions you might have.

**Reply by AlexTMjugador** [reply]
Thanks a bunch for the detailed writeup @Losses, it's much appreciated!

According to your descriptions, your file indeed has a quite unusual initial packet. As I didn't consider such a perversion to the Vorbis specification while developing OptiVorbis, its Vorbis decoding logic is powered by a relatively simple state machine, where state transitions are made in the order header packets are found, and expect such an order to be respected. Changing the initial state of such state machine could be an easy fix, but at the same time there are reasons to think it might not be that easy, given that OptiVorbis spew some errors coming directly from the `ogg` crate. Would you mind sharing a troublesome Ogg Vorbis file for further inspection?

(By the way, I have seen quite a few Ogg Vorbis files in the wild where vendor strings are not encoded in UTF-8 like the specification mandates and you expect [here](https://github.com/Web-Media-Foundation/infrastructure/blob/3a52aa92e596151207631afd8fc047db051b1726/packages/ogg-polyfill/src/OggVorbisPage.ts#L323-L326), but as Windows wide strings and even in ASCII codepages that are not forwards-compatible with UTF-8. I figured this might be an interesting observation to share given your technically interesting collection of Ogg Vorbis files :wink:)

**Reply by Losses** [reply]
> Thanks a bunch for the detailed writeup @Losses, it's much appreciated!
> 
> According to your descriptions, your file indeed has a quite unusual initial packet. As I didn't consider such a perversion to the Vorbis specification while developing OptiVorbis, its Vorbis decoding logic is powered by a relatively simple state machine, where state transitions are made in the order header packets are found, and expect such an order to be respected. Changing the initial state of such state machine could be an easy fix, but at the same time there are reasons to think it might not be that easy, given that OptiVorbis spew some errors coming directly from the `ogg` crate. Would you mind sharing a troublesome Ogg Vorbis file for further inspection?
> 
> (By the way, I have seen quite a few Ogg Vorbis files in the wild where vendor strings are not encoded in UTF-8 like the specification mandates and you expect [here](https://github.com/Web-Media-Foundation/infrastructure/blob/3a52aa92e596151207631afd8fc047db051b1726/packages/ogg-polyfill/src/OggVorbisPage.ts#L323-L326), but as Windows wide strings and even in ASCII codepages that are not forwards-compatible with UTF-8. I figured this might be an interesting observation to share given your technically interesting collection of Ogg Vorbis files 😉)

Hi!

The original file is sus in terms of copyright, so it's unlikely that I can share the original file with you directly. However, I can construct a file with the same structure and provide it to you; this process is not very difficult. I will do it this weekend.

By the way, I also implemented a Rust library specifically for debugging the OGG/Vorbis formats (Heavily helped by GPT, as I don't know how to write Rust myself): https://github.com/Losses/ogg-magic. It can output some detailed information, and this implementation is more useful than others I've encountered before (for debugging purposes). Hope this could help :P

Cheers

**Reply by AlexTMjugador** [reply]
> The original file is sus in terms of copyright, so it's unlikely that I can share the original file with you directly. However, I can construct a file with the same structure and provide it to you; this process is not very difficult. I will do it this weekend.

Sounds good! It'd be great if you could do that indeed :pray: 

> By the way, I also implemented a Rust library specifically for debugging the OGG/Vorbis formats (Heavily helped by GPT, as I don't know how to write Rust myself): https://github.com/Losses/ogg-magic. It can output some detailed information, and this implementation is more useful than others I've encountered before (for debugging purposes). Hope this could help :P

Nice work! I can see that library being very useful for these kind of things. Maybe I end up adding it to the list of related tools in the README :+1:
